### ISSUE TITLE
Contact Names + Photos missing for SMS messages

### ISSUE THREAD

**Issue by derekderaps** [yakyak/yakyak#232]
When I get SMS messages in my Hangouts, the name and photo of the contact don't show up (they do in the regular Hangouts app). It's not a broken image link, just the regular hangout icon,

![default hangouts icon](https://deraps.s3.amazonaws.com/f/photo.jpg)

Any thoughts? Thanks!

**Comment by framerate** [comment]
Yeah, the name just shows as the phone number. Is this an API restriction?

**Comment by davibe** [comment]
I guess messages in this case have a different payload.. never looked as i don't have friends sending sms to me on google hangout

**Comment by panoptican** [comment]
Might be a Google Voice-specific feature now that that service can be integrated in Hangouts. To test, would suggest setting up a free Google Voice account and sending a few SMS to that number from your own phone. I'll probably give fixing it a try later this evening too.

**Comment by adriangb** [comment]
I use Google Voice. Definitely a problem with SMS sent to the GV number.

**Comment by davesimon** [comment]
I can verify this as an issue. I’ll look into the regular Hangouts web app to see what XHR requests it’s making. I’m using a Google Voice number, which for a long time the Hangouts app had the same trouble with. I ended up re-inputting many people into Contacts to get it to work.

**Comment by jacobmischka** [comment]
Same here, though I'm using Project Fi with a ported Google Voice number. Though I'm pretty sure the backend behavior of Fi and regular GV are the exact same, just commenting for completeness's sake.

**Comment by mattrock5a** [comment]
BUMP... would love to see this fix implemented, and know many others who say the same!

**Comment by underscorephil** [comment]
+1

**Comment by bwiernik** [comment]
Also, messages sent to/from the GV number don't display the "SMS" tag over the contact picture.

**Comment by dinoboy197** [comment]
Confirmed to still be an issue in 1.3.3. This is a weakness over Google's official client.

**Comment by AndrewJDR** [comment]
Is this a restriction of hangupsjs? Should an issue be logged there as well?

**Comment by averissimo** [comment]
**We need someone that uses hangouts with sms to debug and find why**

it could be a problem with the library, or maybe the name and photo have a different structure with this type of contact

if someone wants to pick this up I can help showing in which data structures to look

**Comment by AndrewJDR** [comment]
@averissimo Sure, if you want to tell me what to look for (what exact commands to run, what data to look for, etc) I'll try to help.

**Comment by averissimo** [comment]
@AndrewJDR this should require having a development version of yakyak and some skills to pursue stuff if this first approach does not do it :)

__First__ go to `src/ui/views/convlist.coffee` and add the line in bold below line 38, using the same indentation as the line above. 

__`console.log 'Drawing ' + nameof(entity[p.id]) + '\'s avatar', entity[p.id]`__

This is how it should look (with a lot more space in the begging

![image](https://cloud.githubusercontent.com/assets/211358/21197491/92abcfb4-c233-11e6-9c21-9c710db70964.png)

__Second__ compile and run yakyak (`npm start`) and you'll see a lot of new lines in the console, look for the one that should be drawing your contact. Expand the object  on the right _(except the _proto properties)_ and screenshot it here, such as:

![image](https://cloud.githubusercontent.com/assets/211358/21197267/bfbe06da-c232-11e6-92f0-fa52307a0846.png)

**Comment by AndrewJDR** [comment]
@averissimo I sent you an email privately with the screenshots.

I would say the news is not good -- the display name is not contained in any of the members that I can see.

Furthermore, I wanted to mention an observation I made:
When loading the contact list in google hangouts using gmail, I noticed that the contact name doesn't show up immediately. First, it shows the phone number for a fraction of a second, THEN fills in the contact name.

This told me that it's probably performing a separate lookup to obtain the contact name. Realizing this, I tried the following:
* I found an SMS contact in my google hangouts list that did *not* already have a corresponding entry in my google contacts (By "google contacts", I mean https://contacts.google.com) and therefore only showed up as a phone number in my gmail hangouts list. Let's say this contact had the phone number +1-888-888-8888.
* Using https://contacts.google.com, I added a new contact "John Doe" that corresponded with that phone number "+1-888-888-8888"
* I reloaded my gmail. As expected, now showed "John Doe" instead of 888-888-8888. Also as expected, upon reloading gmail, there was a slight delay before the 888-888-8888 was changed to "John Doe" in the contact list. Again, this indicates that it's doing some kind of separate lookup against the google contacts list to fill in this name.

Based on all of this, I believe a separate call (or a set of separate calls) to the google contacts api will be necessary to retrieve the phone number -> Full Name mappings.

**Comment by AndrewJDR** [comment]
Google contact api info is here:
https://developers.google.com/google-apps/contacts/v3/
I'm in a timecrunch now so code-wise, I won't be embarking on this project myself, but I *can* continue to provide data/information, so let me know if I can help you in that department.

**Comment by AndrewJDR** [comment]
Btw, something like this could potentially be used - https://github.com/olalonde/Google-Contacts

**Comment by averissimo** [comment]
Using the google contacts api could be a nice alternative, if the same token could be used for both hangupsjs and gcontacts. Otherwise it could be a pain in the ass :)

Can you check 2 more things?

1. test if tdryer/hangups can recognize the contact's name _(hangups was based on this library -- and it is more heavily developed that ours)_
    * install and run the client to see how that contact appears
    * you can run with your yakyak token: `hangups --token ~/.config/yakyak/refreshtoken.txt` or equivalent directory in you OS
1. there is also some information on the participant in the conversation (see diff and screenshot below)
    * although I guess the fallback name would be the number again

```
+++ src/ui/models/viewstate.coffee
@@ -81,6 +81,7 @@ module.exports = exp = {
         @setLastKeyDown 0
         updated 'viewstate'
         updated 'switchConv'
+        console.log 'Selected new conversation', conv[conv_id]
```

![image](https://cloud.githubusercontent.com/assets/211358/21233583/959e0dd6-c2e7-11e6-85bc-4ec01ac264c6.png)

This is what tdryer/hangups shows at my test account, if number appears here then we'll probably have to implement a call to contacts api

![outplpxcerqoo](https://cloud.githubusercontent.com/assets/211358/21233512/49619294-c2e7-11e6-9194-08b36ab5005c.gif)

**Comment by AndrewJDR** [comment]
@averissimo 
1. Regarding hangups, unfortunately, I tried with the latest git of hangups and it shows only the number for that contact. I used this release - https://github.com/tdryer/hangups/tree/7ec7a73c7f85159807424b53ddbbeea4524900ba
I'll email you a couple of screenshots just for your reference.

2. At first I got excited when I saw a name here, but I then realized that this data structure shows both participants, and one of them is me. So I had a bit of false hope there. Ultimately, the participant_data doesn't seem to contain any name information about the contact. I've again emailed you a screenshot of that.

And yeah, i was wondering the same thing about the authentication token -- I'm not sure whether or not it can be shared between the contacts api and hangouts.

**Comment by averissimo** [comment]
Your initial assesment was correct, it was a hopeful approach though ;)

Our current token does not ask for this, I tried the library you referenced with current token

If we introduce the contacts API users would have to logout and login again to get a new token that adds that authorization

To avoid forcing all users having to do this, I'd guess a non-obstrusive message each time an SMS account is detected could be presented hinting the user to logout->login

But that would be something that who wants to pick this up have to decide.

**Comment by averissimo** [comment]
btw on my test of that library, I could not find any mapping from `chat_id` and `contact_id` :\

**Comment by AndrewJDR** [comment]
@averissimo Hmm, I guess that doesn't really surprise me so I'm not disappointed.

I always assumed it would need to map using the google contact phone number <-> fallback_id or similar (both stripped of anything but the numbers), not anything like chat_id or chat_id?

**Comment by hadees** [comment]
@averissimo this is a feature I really want, i'd be willing to create a service class or something that handles the mapping.  It seems like Hangouts is the best SMS for Google Fi and YakYak seems like the best OSX app for hangouts.

**Comment by AndrewJDR** [comment]
@hadees If you look at the google people api: https://github.com/google/google-api-nodejs-client/blob/master/apis/people/v1.js it should be possible to pull the contacts using that, then cross-reference these contact's phone number's against the fallback_name on the hangouts side.

This guide looks to give a decent start on using the people api:
https://developers.google.com/people/quickstart/js
And this project has the client api in js:
https://github.com/google/google-api-nodejs-client
This is the call to get all the user's contacts:
https://developers.google.com/people/api/rest/v1/people.connections/list

I think the slightly tricky business about all this is that the existing hangouts oauth token is no good for doing queries against the people api -- it doesn't have the proper scopes in place. When you try to use the people api with the existing hangupsjs refresh token, you will probably get an error ("You must add People API access to this project, log into the project on google and add it", or similar). Problem is, you do not have the ability to edit that project in order to add people api access to it (by "that project", i mean https://github.com/yakyak/hangupsjs/blob/master/src/auth.coffee#L12). To overcome this. I believe you may have to create a completely separate project, add google people api access to that, then when the user logs into yakyak, obtain and separately save an oauth refresh token against that project so you can make queries from the user's contact list.

**Comment by averissimo** [comment]
@AndrewJDR first of all, very informative comment \_o\_ 

Second, both scopes can live under same oauth token, right? if so I believe we can just invalidate existing one and force user to re-login once this goes live.

@hadees once you are able to get the mapping via (some API) I can definitely help you integrate with YakYak

on getting the right api, well.. I guess that will require some try and fail approach :) with people api or other.

**Comment by AndrewJDR** [comment]
@averissimo -
> Second, both scopes can live under same oauth token, right? if so I believe we can just invalidate existing one and force user to re-login once this goes live.

Unfortunately, I don't think this will work. Even if you add the google people scope to the existing oauth request on [this line](url), the moment you try to make a call on the people api, you'll get an error like "You must add People API access to this project, please log into the project on some google url and add it", or similar.

Basically, it seems that you need both the scope in the oauth request AND you have to add the google people api permission on the google side using the project website. And again, we don't have access to this project's website, since the project is "Hangouts for iOS" or whatever -- https://github.com/tdryer/hangups/blob/master/hangups/auth.py#L4

**Comment by hadees** [comment]
what about just using the local contacts?  If they are already syncing with google.

**Comment by jonmartin345** [comment]
Just wanted to chime in, I'm a Mac user, and would love this feature to become available.

My knowledge of web/app development is limited, but I'm available to test/debug if it would help.

Thanks for your awesome work, love the app.

**Comment by dcrbeloit** [comment]
Sorry new to this,  My understanding is that Google voice is intergreated now for SMS,  but the contact information is still missing is that correct?

**Comment by nikoargo** [comment]
Yup! Just tried it out.

**Comment by davidlmobley** [comment]
I'm trying this out now and use SMS via Google Voice routinely (primarily!) and can verify the contact info is still missing. I can perhaps help a small amount if needed as I have some minimal programming abilities.

**Comment by t1nypotato** [comment]
I can verify that the contact information is missing as well on 1.4.4.

**Comment by averissimo** [comment]
This could be a middle ground solution (before using an actual API)

https://github.com/tdryer/hangups/pull/200

Can someone check out if any object contains the user's name?

My suggestion on how to do it:

1. Select a sms conversation
1. Open debugger pressing ctrl + shift + i
1. Go to debugger tab
1. Press Ctrl + P
1. Write messages.coffee and open that file
1. Create a (debugging) breakpoint at around `if c?.current_participant?`
1. Switch to yakyak and open a sms conversation or just wait until debugger kicks in
1. Inspect the following objects:
    - `c.participant_data`
    - `entity[c.current_participant[0].chat_id]`
    - `entity[c.current_participant[1].chat_id]`

Let us know if there is a name in any of those objects *(please respect people privacy by not showing ids/names/phone numbers/email/etc..)*

**Comment by pandeyharsh** [comment]
@averissimo I'm very new to this and I am willing to help if possible. Where do I execute step 1 to 6? Do I have to install 'hangups' separately?

**Comment by averissimo** [comment]
This should all be done inside YakYak

**Comment by pandeyharsh** [comment]
I have left part of the `chat_id` visible so that you can verify that I have posted the correct objects.

<img width="322" alt="screen shot 2018-01-23 at 2 13 42 pm" src="https://user-images.githubusercontent.com/14907899/35298849-0eb318f2-0049-11e8-911f-bdf828323383.png">


`entity[c.current_participant[1].chat_id]` contains my name but the other object does not contain the name of the other participant.

<img width="302" alt="screen shot 2018-01-23 at 2 13 15 pm" src="https://user-images.githubusercontent.com/14907899/35298861-13dd590a-0049-11e8-8632-7976f105d9b5.png">

Not sure if it matters, but I'm a Project Fi user, and I ported my number from Google Voice.

**Comment by hazel-nut** [comment]
@averissimo, i can confirm what @pandeyharsh saw. i see my own name in `c_participant_data[1]` and `entity[c.current_participant[1].chat_id]`, but no other names.

**Comment by stealthybox** [comment]
Hello. This thread encouraged me to try YakYak.

I'm a Project Fi + Hangouts user and I ended up here, because Google is deprecating the Chrome App. The Extension doesn't work for my auth needs right now.

I installed YakYak v1.5.1 with `brew cask install yakyak`.

**My contact names are showing up for some people 🙂**
Names show up in group MMS messages as well for those same individuals.

Google may have updated their API responses, or we may have merged code changes that fixes some of this.
It's notable that my extensive use of Hangouts on my phone may be contributing to more metadata being available. I've noticed laggy contact name behavior when adding new contacts with my phone, even with the official Google apps on both my laptop and phone.

Contact photo's don't show up for SMS in YakYak.
There is no SMS badge or phone number that is displayed when the names do show up.
It's hard to differentiate an SMS vs the Hangout user's chat for the same person when the name shows up.

I'm able to send and recieve messages for an existing SMS conversation.
When adding a new conversation, the "Search people" field doesn't seem to search my Phone contacts.
I haven't tried just putting in a phone number and seeing if that works to send an SMS yet.

I'll report back as I find bugs / missing functionality.

**Comment by stealthybox** [comment]
I can't seem to send/upload a photo via text.
It just spins:
```
uploading image… <donut_large>
```

**Comment by slvtrs** [comment]
Great work so far! I'd like to pick this up and get a proper contact sync working if possible.
@stealthybox @averissimo @AndrewJDR @pandeyharsh @hazel-nut 

My situation may be unique in that I don't us Fi, but a Voice number set to "receive ssm in Hangouts".
Although I can also confirm that no `participant_data` or `entity` objects contain names, not even my own.

If the existing oauth doesn't have proper scoping to access the google people api, perhaps a user could create their own google api project with access to people and connect it to yakyak upon sign in?

**Comment by stealthybox** [comment]
@slvtrs do you have a large volume of SMS conversations that are missing names?
I'm not positive what the differentiating factor is between my contacts that do and do not have name data.

I was able to determine that none of my MMS Group chats have any name data backed into them.
However, when somebody in that MMS Group chat has a 1-on-1 SMS conversation with me and they have a name, YakYak is able to fetch their name from the central store using their chatid and render it in the Group chat.

**Comment by oakbrad** [comment]
I am a Project Fi subscriber who exclusively uses Yakyak for SMS, it seems like about 90% of my conversations do not have contact names yet a handful do. @stealthybox I think you might be right, it does appear that it loads contact names if you have been in a group chat with them.

I have a colleague I've only texted previously in a group chat, and yesterday texted him directly for the first time and was surprised to see Yakyak pull the name. The name matches what I have in my Google Contacts (not his actual name but the alias I entered).

Happy to help debug as I think Yakyak is really the only way to SMS on Linux desktop and this is a pretty crucial feature, I'm afraid I'm going to text the wrong person :).

**Comment by oojacoboo** [comment]
*bump* 2 year old issue that's critical.  What good is a chat app where you have to reference contacts when they show up with some random phone number?  The app is useless without this issue resolved.  Also, doing a Google Person API sync and cross referencing isn't very difficult.

**Comment by averissimo** [comment]
Again, we need a developer from the USA *(or ?Canada?)* that is able to work on improving the sms capabilities of Yakyak.

**Comment by stealthybox** [comment]
@oojacoboo you might try a more constructive/positive approach in the future to convey the importance of a feature.

This app is free and open source, and it's really cool.
It's maintained by people in their free time.
I don't connect at all with the sentiment of your comment.

If you see some easy things that can be done to improve this area of the app, please do contribute a design or some code.

I believe the major blocker on the contacts api is the Oauth scopes the app currently has access to.

**Comment by chambery** [comment]
Would it be possible to use a local hash for these lookups?  I'd be content to just enter the 20 or so names-for-numbers in the specific instance of Yak Yak, even if I have to do for multiple installations. 

Even better if it were possible to import from a file.

**Comment by stealthybox** [comment]
@chambery when messing around in the electron dev-console, I could see that YakYak did use a central store for this information.
This is why some of my MMS threads have names/pictures and some do not. (The store is being updated from another SMS thread, so the MMS UI can associate the name to the number.)

From what I can tell, this local hash structure already exists, and we'd just need the UI bits and a little logic to patch it properly.

**Comment by willyli2319** [comment]
+1 Fi user here as well. I'm a CS student. I do not have any experience in js, but I'll see what I can do with the resources linked by @AndrewJDR

**Comment by copperly123** [comment]
+1 on v1.5.3 Linux

**Comment by lime45** [comment]
+1 Fi user here. v1.5.4 Linux

**Comment by Kommynct** [comment]
As a temporary workaround, perhaps add a way to rename contacts just within yakyak?

obviously it'd be better for it to be handled automatically, but if people were renamable/you could change their photo to what you'd like, if you're willing to set that up, it could be better than an automatic system.

**Comment by StevenSeiden** [comment]
I agree that "nicknaming" people would be an acceptable solution for the time being.

**Comment by copperly123** [comment]
I agree as well

**Comment by emuehlstein** [comment]
+1 on nicknaming.  Google Fi, 1.5.3 mac

**Comment by JohnnyTToxic** [comment]
+1 on nicknaming as well, some simple local config containing the SMS number pair with "friendly name" would be great.

**Comment by LudusLight** [comment]
+1 on nicknaming, please. This is a huge issue that is preventing me from using this application currently.

**Comment by cbartondock** [comment]
I have some experience developing electron apps, albeit with typescript, and am a project fi user in the US that would love to have this feature working. I'll look fork it and see what I can do.

**Comment by cbartondock** [comment]
It seems the People API is the way to go (I don't think yak yak needs the full contacts api which gives the ability to add and edit contacts). In the Google dev console are we currently using the deprecated Google + Hangouts API or the Hangouts Chat API?

**Comment by cbartondock** [comment]
Okay forked hangups. It might be possible to just ammend the `OATH2_SCOPES` there and get authorization for the People API

**Comment by cbartondock** [comment]
Probably the better solution is indeed a second login whenever any "sms" conversations are detected, linking to a wholly separate google project that has the People API enabled. ugh
